org.aposin.licensescout.core/src/main/java/org/aposin/licensescout/finder/JavaJarFinder.java

```diff
@@ -241,25 +225,49 @@ private void parsePackedJarArchive(final Archive archive, final InputStream file
                         addMessageDigest(finderHandler, entryContainer, foundArchive);
                     }
                     addLicenseFromManifest(foundArchive, archiveMetaInformation, newFilePath2);
-                    parsePackedJarArchive(foundArchive, entryContainer.getInputStream(), file, newFilePath2);
+                    try (final InputStream inputStream = entryContainer.getInputStream()) {
+                        parsePackedJarArchive(foundArchive, inputStream, file, newFilePath2);
+                    }
                 } else {
                     if (finderHandler.isFile(entry)) {
-                        if (isCandidateLicenseFile(entryName)) {
-                            archive.addLicenseCandidateFile(newFilePath);
-                        }
-                        final Collection<License> licenses = checkFileForLicenses(jarInputStream, entryName,
-                                getLicenseStoreData());
-                        addLicenses(archive, licenses, newFilePath);
                         final JarEntryContainer entryContainer = finderHandler.createEntryContainer(jarInputStream);
-                        if (isPomFile(entryName)) {
-                            addLicensesFromPom(entryContainer, archive, newFilePath);
-                        }
+                        handleArchiveNormalFile(archive, entryName, entryContainer, newFilePath);
                     }
                 }
             }
         }
     }
 
+    private void handleArchiveFile(final String entryName, final FileSystemEntryContainer entryContainer,
+                                   final String filePath)
+            throws IOException, Exception {
+        final FinderHandler<File, FileSystemEntryContainer, File> finderHandler = fileSystemFinderHandler;
+        final ArchiveMetaInformation archiveMetaInformation = finderHandler
+                .getArchiveMetaInformationFromManifest(entryContainer);
+        final Archive foundArchive = createAndAddArchive(entryName, archiveMetaInformation, filePath);
+        addMessageDigest(finderHandler, entryContainer, foundArchive);
+        final String newFilePath = filePath + "!";
+        addLicenseFromManifest(foundArchive, archiveMetaInformation, newFilePath);
+        try (final FileInputStream archiveFileInputStream = new FileInputStream(entryContainer.getFile())) {
+            parsePackedJarArchive(foundArchive, archiveFileInputStream, entryContainer.getFile(), newFilePath);
+        }
+    }
+
+    private void handleArchiveNormalFile(final Archive archive, final String entryName,
```

Comment:
[OPTIONAL] would be clearer to call this method like "handleUnpackedArchive" or "handleExtractedArchive" or something similar instead of "NormalFile". What do you think?

Reply:
There are methods 'parseUnpackedJarArchive' and 'parsePackedJarArchive' (and also 'parseFile'). These are the methods that have a loop over their contained file entries (in a directory), respective JarEntries in a JAR. The files/entries are checked if they are archives. For these, 'handleArchiveFile' is called, which goes into recursion to 'parsePackedJarArchive'.
If a file is not an archive, it is considered a 'normal file', which may be a README, License file, about file, POM or something else that contains meta information. This is done in 'handleArchiveNormalFile'. Note that this method is called both from parsePackedJarArchive and parseUnpackedJarArchive, so it is not specific to unpacked archives. Your method name suggestions would maybe apply to the method 'handleArchiveFile', which actually is used only for unpacked archives (called from parseFile and parseUnpackedJarArchive. (By the way, parseFile should rather be named 'parseDirectory') The method 'handleArchiveNormalFile' should keep the 'normalFile' and could maybe drop 'Archive', because the file handled is only IN an archive (which is more or less self-evident) but IS not an archive itself.

Reply:
Had another look: 'parseFile' handles indeed both a file and a directory case and calls itself in case of a directory for each directory entry. So contrary to the statement above, it is not only for directories. One could argue that the recursive calling structure could still be improved. However, for now I have no really good idea. This is actually a step-by-step process, and I like to have one step stable in the repository before doing the next experiment.